tools/walletextension/subscriptions/subscriptions.go

```diff
@@ -0,0 +1,131 @@
+	return &SubscriptionManager{
+		logger: logger,
+	}
+}
+
+func (s *SubscriptionManager) HandleNewSubscriptions(clients []rpc.Client, req *wecommon.RPCRequest, resp *interface{}, userConn userconn.UserConn) error {
+	if len(req.Params) == 0 {
+		return fmt.Errorf("could not subscribe as no subscription namespace was provided")
+	}
+
+	// create a chanel that will collect the data from all subscriptions
+	commonChannel := make(chan common.IDAndLog)
+
+	// save subscriptions
+	subscriptions := make([]*gethrpc.ClientSubscription, 0, len(clients))
+	subscriptionIDS := make([]string, 0, len(clients))
+
+	// TODO: Create a buffered channel and perform deduplication of logs or implement additional logic to filter logs
+
+	// Do periodic checks if userConn is closed and unsubscribe from all subscriptions for this user
+	go func() {
+		for {
+			if userConn.IsClosed() {
+				for _, subscription := range subscriptions {
+					subscription.Unsubscribe()
+				}
+				subscriptions = []*gethrpc.ClientSubscription{}
+				return
+			}
+			time.Sleep(100 * time.Millisecond)
+		}
+	}()
+
+	// Send all logs from common channel to user (via userConn)
+	go func() {
+		for idAndLog := range commonChannel {
+			if userConn.IsClosed() {
+				// Log that websocket was closed - unsubscribing is handled by periodic checks in separate goroutine
+				s.logger.Info("received log but websocket was closed on subscription", log.SubIDKey, idAndLog.SubID)
+				return
+			}
+
+			jsonResponse, err := wecommon.PrepareLogResponse(idAndLog)
+			if err != nil {
+				s.logger.Error("could not marshal log response to JSON on subscription.", log.SubIDKey, idAndLog.SubID, log.ErrKey, err)
+				continue
+			}
+
+			err = userConn.WriteResponse(jsonResponse)
+			if err != nil {
+				s.logger.Error("could not write the JSON log to the websocket on subscription %", log.SubIDKey, idAndLog.SubID, log.ErrKey, err)
+				continue
+			}
+		}
+	}()
+
+	// loop over all clients and create a new subscription for each of them
+	s.logger.Info(fmt.Sprintf("Subscribing to: %d clients", len(clients)))
+	for _, client := range clients {
+		s.logger.Info(fmt.Sprintf("Subscribing for an event with client: %s", client))
+		var subscriptionID interface{}
+		subscription, err := s.addSubscription(client, req, &subscriptionID, commonChannel)
+		if err != nil {
+			s.logger.Info(fmt.Sprintf("Error: %v", err))
+		} else {
+			// If there was no error, the subscription was successful. Store it for unsubscribing in the future
+			subscriptions = append(subscriptions, subscription)
+
+			// check if subscriptionID is of type string
+			if strSubscriptionID, ok := subscriptionID.(string); ok {
+				subscriptionIDS = append(subscriptionIDS, strSubscriptionID)
+				s.logger.Info(fmt.Sprintf("Subscribed with subscription ID: %s", subscriptionID))
+			} else {
+				s.logger.Info(fmt.Sprintf("Error: invalid type of subscriptionID. Expected string, got: %T", subscriptionID))
+			}
```

Comment:
Won't this mean the subscriptions has appended the subscription ?

What if this `check if subscriptionID is of type string` is done at the `err != nil` level ?

Reply:
fixed